src/core/analysis/mcmc/Mcmcmc.cpp

```diff
@@ -1868,15 +1868,38 @@ void Mcmcmc::tune( void )
         // if the heat of a given hot chain is smaller than the minimum bound
         // interpolate this heat and the heats of all the hotter chains
         // to fall between the lowest heat that is greater than the minimum bound and the minimum bound
-        if (j < num_chains)
+        
+        std::vector<double> greaterThanMin;
+        std::vector<size_t> badChainIdx;
+        
+        // std::cout << "The following should be sorted from largest to smallest:" << std::endl;
+        for (size_t l = 0; l < num_chains; ++l)
         {
-            double rho = pow(chain_heats[colderChainIdx] / heatMinBound, 1.0 / (num_chains - j));
-            size_t k = j;
+            // std::cout << chain_heats[ chainForHeatIndex(l) ] << std::endl;
+            if ( chain_heats[ chainForHeatIndex(l) ] > heatMinBound )
+            {
+                greaterThanMin.push_back( chain_heats[ chainForHeatIndex(l) ] );
+            }
+            else
+            {
+                badChainIdx.push_back(l);
+            }
+        }
+        
+        // std::cout << "We have " << badChainIdx.size() << " bad chain(s)" << std::endl;
+        if ( badChainIdx.size() != 0 )
+        {
+            double lowestGreaterThanMin = *std::min_element( greaterThanMin.begin(), greaterThanMin.end() );
```

Comment:
does that work if all chains are bad ? is that a possibility ?

Reply:
Do we know which operation produce the NaN, and what its arguments where?

Also, I am trying to follow the interpolation math.  It seems a lot easier on the log scale.  Could we add a 1-line comment?

Reply:
> Do we know which operation produce the NaN, and what its arguments where?

Yes – I described it in detail in a [comment](https://github.com/revbayes/revbayes/issues/651#issuecomment-2642082750) on the original issue, but probably should have done it here instead. The `nan` was coming from the `rho` constant, because when the initially proposed value of `chain_heats[colderChainIdx]` was negative (which can occasionally happen), we ended up trying to take the square root of a negative number.

Adding a comment is a good idea; will do.

Reply:
```
            // We want to linearly interpolate the log(heat) between (j-1, log(chain_heats[colderChainidx])) and (num_chains-1, log(heatMinBound))
            // The formula for linearly interpolating between (x1,y1) and (x2,y2) is y(x) = y1 + (y2-y1)*[ (x-x1)/(x2-x1) ]
            //
            // Therefore log(chain_heats[k]) = log(chain_heats[colderChainIdx]) + (log(heatMinBound) - log(chain_heats[colderChainIdx]))
            //                                         * (k - (j-1))/((num_chains-1) - (j-1))

```

Reply:
We could also implement it this way.

Reply:
Have you actually tested that on `950_ns_ki`? I'm not convinced the original code with `chain_heats[colderChainidx]` was always guaranteed to pick the lowest heat greater than the minimum.

Reply:
If we want to avoid having the hottest chain heat equal to `heatMinBound`, we can accomplish this simply by changing `x2` to `num_chains`.

Hmm.... interesting that the `chain_heats[colderChainIdx]` could be negative.  That's unfortunate.